misc/helper-scripts/rosmsg_to_asn1.py

```diff
@@ -0,0 +1,417 @@
+        lib_asn = composed_type[1]+'-Types'
+        # libraries[asn1_type] = lib_asn
+        primitive = False
+    return asn1_type, lib_asn, primitive
+
+
+def generate_libraries_string(import_libraries):
+    """
+    Generate a string from a set of libraries to be imported to the ASN.1 type.
+    :param import_libraries: The libraries to be imported.
+    :return s: The libraries names in string format.
+    """
+    s = "IMPORTS "
+    comma = ""
+    for key in import_libraries:
+        num_types = len(import_libraries[key])
+        for t in import_libraries[key]:
+            s += comma + t
+            comma = ", "
+        s += " FROM " + key + " "
+        comma = ""
+    s += ";"
+    s += "\n"
+    return s
+
+
+def process_msg(rospack, msg):
+    """
+    Process a message creating a string containing an ASN.1 representation.
+    :param rospack: A rospkg instance.
+    :param msg: The ROS message to be converted to ASN.1 message.
+    :return msg_string: The string containing the ASN.1 representation.
+    :return msg_deps: Any message dependencies that need to be generated.
+    """
+    msg_name = None
+    pkg_name = None
+    msg_string = None
+    msg_deps = None
+    # Get the package contining the message
+    msg_name, pkg_name = get_message_package(rospack, msg)
+    if msg_name is None or pkg_name is None:
+        return msg_string, msg_deps
+    # Get the Python module contining the package's messages
+    module = get_python_msg_module(pkg_name)
+    if module is None:
+        return msg_string, msg_deps
+
+    # Get the message as a dictionary
+    msg_dict = get_message_dict(module, msg_name)
+    if msg_dict is None:
+        return msg_string, msg_deps
+
+    msg_string = (msg_name + "-Types DEFINITIONS ::=\nBEGIN\n")
+    msg_deps = []
+    slot_types = msg_dict["slot_types"]
+    slots = msg_dict["slots"]
+    num_fields = len(slot_types)
+
+    import_libraries = dict()
+
+    slot_asn1_types = []
+    field_types = []
+    lengths = []
+
+    for slot_type in slot_types:
+        slot_type, field_type, length = find_field_type(slot_type)
+        asn1_type, lib_asn, primitive = find_ASN1_type(slot_type)
+        if lib_asn in import_libraries:
+            if asn1_type not in import_libraries[lib_asn]:
+                import_libraries[lib_asn].append(asn1_type)
+        else:
+            import_libraries[lib_asn] = [asn1_type]
+        slot_asn1_types.append(asn1_type)
+        field_types.append(field_type)
+        lengths.append(length)
+        if not primitive:
+            msg_deps.append(slot_type)
+
+    msg_string += generate_libraries_string(import_libraries)
+    fixed_indices = [i for i, x in enumerate(field_types)
+                     if x == "fixed-length"]
+    variable_indices = [i for i, x in enumerate(field_types)
+                        if x == "variable-length"]
+
+    for i in fixed_indices:
+        msg_string += "L" + slots[i].replace("_", "-") + "::="
```

Comment:
Would it make sense to do this replacement in a list-comprehension at initialization of slots[]? ie.

```
slots = [s.replace("_","-") for s in msg_dict["slots"]]
```

Reply:
Yes you save some replace that way I suspect. Fixed